### PR TITLE
Add `minetest.settings` to CSM API and allow CSMs to provide `settingtypes.txt`

### LINKED ISSUE #10382: Add minetest.settings to CSM API
The code is already written as part of https://github.com/EliasFleckenstein03/dragonfireclient. This would allow changing client settings without restarting the game using a CSM.

### PR DESCRIPTION
**Pull request by AFCMS**
Add compact, short information about your PR for easier understanding:

Fix #10382
Allow CSMs to use `minetest.settings`.
Allow CSMs to provide `settingtypes.txt`

## To do

This PR is Ready for Review.

- [x] Expose `minetest.settings` to CSM API
- [x] Expose `minetest.setting_get_pos` to CSM API
- [x] Add test to preview CSM
- [x] Allow client mods to provide `settingtypes.txt`
- [x] Add documentation

## How to test

Use this simple script inside a CSM and check the console output:
```lua
print(minetest.settings:get_bool("enable_client_modding", false))
```

Open the settings tab and check is there is a `Client Mods` tab.

### CODE REVIEW

**Comment by rubenwardy** [comment]
This has already been rejected in #5454

**Comment by TurkeyMcMac** [comment]
Also, if settings support is added and is later found to be a bad idea for whatever reason, it can just be removed, since the CSM API is unstable.

**Comment by AFCMS** [comment]
> > especially when the features are potentially security sensitive like exposing the entire settings object (which may contain secrets)
> 
> That's indeed critical, and this can be treated as a whitelist/blacklist of settings that can be safely exposed or not, or only expose an object key ("safe"?), where the publicly safe settings can live, so there would no need to maintain any list.

As I said above, CSMs are user installed.
It doesn't make sense for SSMs to have access to all your settings but not CSMs.

**Comment by rubenwardy** [comment]
The CSM API and sandbox is designed for SSCSMs, which is why it's so restricted

**Comment by rubenwardy** [comment]
The current plan is to execute CSMs and SSCSMs in the same sandbox

**Comment by sfan5** [comment]
> The current plan is to execute CPCSMs and SSCSMs in the same sandbox

This is not feasible.
To ease implementation during the transition period CPCSM could be automatically disabled when SSCSMs are in use but if the final goal is to have both then they must be in separate environments.

**Comment by AFCMS** [comment]
I just added the `client_lua_api.txt` entry for `minetest.settings`

**Comment by AFCMS** [comment]
I just exposed `minetest.setting_get_pos` to the CSM API (by moving its definition to `misc_helpers.lua`)

**Comment by sfan5** [comment]
Approved for inclusion by coredev meeting.

**Comment by appgurueu** [comment]
Needs changes & isn't a possible close anymore

**Comment by AFCMS** [comment]
Rebased

File: builtin/common/misc_helpers.lua [hunk 1]
```diff
@@ -279,6 +279,15 @@ function core.wrap_text(text, max_length, as_table)
 	return as_table and result or table.concat(result, '\n')
 end
 
+
+function core.setting_get_pos(name)
+	local value = core.settings:get(name)
```
Comment (sfan5) [inline builtin/common/misc_helpers.lua:284]: this is not supposed to be here, merging mistake

File: builtin/game/misc.lua [hunk 2]
```diff
@@ -121,6 +121,44 @@ function core.get_player_radius_area(player_name, radius)
...
 
+function core.hash_node_position(pos)
+	return (pos.z + 32768) * 65536 * 65536
+		 + (pos.y + 32768) * 65536
+		 +  pos.x + 32768
+end
+
+
+function core.get_position_from_hash(hash)
+	local x = (hash % 65536) - 32768
```
Comment (sfan5) [inline builtin/game/misc.lua:132]: same here

**Review by sfan5 — CHANGES_REQUESTED** [review]

File: builtin/common/misc_helpers.lua (hunk 1 above)
Comment (AFCMS) [reply builtin/common/misc_helpers.lua:284]: But its the only way to get access to it from the CSM API, right?

Comment (sfan5) [reply builtin/common/misc_helpers.lua:284]: If you have to then duplicate it in `builtin/client/`, not in some random other place.

Comment (AFCMS) [reply builtin/common/misc_helpers.lua]: Is it OK now?

File: builtin/game/misc.lua (hunk 2 above)
Comment (AFCMS) [reply builtin/game/misc.lua]: Fixed?
I removed the duplicated

**Review by sfan5 — APPROVED** [review]
Tested

**Review by SmallJoker — CHANGES_REQUESTED** [review]
Does not protect against writing of `secure.` settings; should be guarded by `checkSettingSecurity`.

**Comment by sfan5** [comment]
relevant code: https://github.com/minetest/minetest/blob/d400a98ef0408ccbd2cc4c72f8b48c3813dfce41/src/script/lua_api/l_settings.cpp#L45-L46

**Comment by AFCMS** [comment]
> Does not protect against writing of `secure.` settings; should be guarded by `checkSettingSecurity`.

But what is exactly to change?
Doesn't `checkSettingSecurity` works for both environments by checking for `secure.`?
Or is it a problem with `ScriptApiSecurity::isSecure()`?

**Comment by SmallJoker** [comment]
```lua
print(dump(minetest.settings:get("secure.http_mods")))
print(dump(minetest.settings:set("secure.http_mods", "test")))
```

The problem is that any CSM could write to `secure.` settings, even disabling mod security for local servers.

**Comment by sfan5** [comment]
The fix is probably this but someone will have to verify it:
```diff
diff --git a/src/script/cpp_api/s_security.cpp b/src/script/cpp_api/s_security.cpp
--- a/src/script/cpp_api/s_security.cpp
+++ b/src/script/cpp_api/s_security.cpp
@@ -417,6 +417,12 @@ void ScriptApiSecurity::setLuaEnv(lua_State *L, int thread)
 
 bool ScriptApiSecurity::isSecure(lua_State *L)
 {
+#ifndef SERVER
+       auto script = ModApiBase::getScriptApiBase(L);
+       // CSM keeps no globals backup but is always secure
+       if (script->getType() == ScriptingType::Client)
+               return true;
+#endif
        lua_rawgeti(L, LUA_REGISTRYINDEX, CUSTOM_RIDX_GLOBALS_BACKUP);
        bool secure = !lua_isnil(L, -1);
        lua_pop(L, 1);
```

**Comment by AFCMS** [comment]
I go tomorrow on vacation and will not able to push/test something for about two weeks. :sweat: 

You can edit the pull request if someone test the fix.

**Review by SmallJoker — APPROVED** [review]
Works.

**Pull request merged**